sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs

```diff
@@ -63,5 +68,61 @@ public TClient GetClient(IServiceProvider serviceProvider, object options, Token
                 return _cachedClient;
             }
         }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_cachedClient != null)
+            {
+                if (_asyncDisposable)
+                {
+                    IAsyncDisposable disposableClient;
+
+                    lock (_cacheLock)
+                    {
+                       if (_cachedClient == null)
+                       {
+                           return;
+                       }
+
+                       disposableClient = (IAsyncDisposable)_cachedClient;
+                       _cachedClient = default;
+                    }
+
+                    await disposableClient.DisposeAsync().ConfigureAwait(false);
+                }
+                else if (_disposable)
+                {
+                    Dispose();
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            if (_cachedClient != null)
+            {
+                if (_disposable)
+                {
+                    IDisposable disposableClient;
+
+                    lock (_cacheLock)
+                    {
+                       if (_cachedClient == null)
+                       {
+                           return;
+                       }
+
+                       disposableClient = (IDisposable)_cachedClient;
+                       _cachedClient = default;
```

Comment:
This one is weird... the compiler doesn't like a `null` assignment here because there's no constraint ensuring `TClient` is a reference type.   But above, it does not like a comparison to `default` for the same reason.